lib/mime.js

```diff
@@ -0,0 +1,126 @@
+'use strict';
+const { parse } = internalBinding('mime_wrap');
+const {
+  Apply,
+  MapIteratorPrototypeNext,
+  RegexpPrototypeTest,
+  StringPrototypeReplace,
+  SymbolReplace,
+  SymbolIterator,
+  TypeError,
+  SafeMap: Map,
+  SafeWeakMap: WeakMap,
+} = require('internal/safe_globals');
+
+const DATA = new WeakMap();
+const $ = (self, guard) => {
+  if (!DATA.has(self)) {
+    throw new TypeError('Invalid receiver');
+  }
+  const data = DATA.get(self);
+  if (guard !== data.guard) {
+    throw new TypeError('Invalid receiver');
+  }
+  return data.value;
+};
+const NotHTTPTokenCodePoint = /[^\u0021\u0023-\u0027\u002A\u002B\u002D\u002E\u005E\u005F\u0060\u007C\u007EA-Za-z0-9]/g;
+NotHTTPTokenCodePoint[SymbolReplace] = null;
+const encode = (value) => {
+  NotHTTPTokenCodePoint.lastIndex = 0;
+  const encode = Apply(RegexpPrototypeTest, NotHTTPTokenCodePoint, [value]);
+  if (!encode) return value;
+  const escaped = Apply(StringPrototypeReplace, value,
+                        [/["\\]/g, '\\$&']);
+  const ret = `"${escaped}"`;
+  return ret;
+};
+class MIMEParams {
+  constructor() {
+    DATA.set(this, {
+      value: new Map(),
+      guard: MIMEParams
+    });
+  }
+  delete(name) {
+    return $(this, MIMEParams).delete(name);
+  }
+  get(name) {
+    const data = $(this, MIMEParams);
+    if (data.has(name)) {
+      return data.get(name);
+    }
+    return null;
+  }
+  has(name) {
+    return $(this, MIMEParams).has(name);
+  }
+  set(name, value) {
+    return $(this, MIMEParams).set(name, value);
+  }
+  *entries() {
+    return yield* $(this, MIMEParams).entries();
+  }
+  *keys() {
+    return yield* $(this, MIMEParams).keys();
+  }
+  *values() {
+    return yield* $(this, MIMEParams).values();
+  }
+  [SymbolIterator]() {
+    $(this, MIMEParams);
+    return this.entries();
+  }
+}
+class MIME {
+  constructor(string) {
+    const data = parse(string, { __proto__: null });
+    const parameters = new MIMEParams();
+    for (var i = 0; i < data.parameters.length; i++) {
+      const [k, v] = data.parameters[i];
+      if (!parameters.has(k)) {
+        parameters.set(k, v);
+      }
```

Comment:
If you’re deduplicating here, wouldn’t it make sense to use a map in C++ rather than a vector of pairs? (e.g. `std::unordered_map<std::string, std::string>`, you can check whether a key already exists by using `map.count(key) == 1`)

Reply:
I'd like to add `.getAll` etc. to match URLSearchParams and the order of the keys is important/preserved per WHATWG spec so map needs to be ordered on insertion and has tests around it.